contracts/bridge/contracts/common/message/InboundLane.sol

```diff
@@ -0,0 +1,301 @@
+        // This should never happen if proofs are correct
+        require(latest_received_nonce <= last_delivered_nonce, "Lane: InvalidReceivedNonce");
+        if (latest_received_nonce > last_confirmed_nonce) {
+            uint256 new_confirmed_nonce = latest_received_nonce;
+            uint256 front = relayersRange.front;
+            uint256 back = relayersRange.back;
+            for (uint256 index = front; index <= back; index++) {
+                UnrewardedRelayer storage entry = relayers[index];
+                if (entry.messages.end <= new_confirmed_nonce) {
+                    // Firstly, remove all of the records where higher nonce <= new confirmed nonce
+                    delete relayers[index];
+                    relayersRange.front = index + 1;
+                } else if (entry.messages.begin < new_confirmed_nonce) {
+                    // Secondly, update the next record with lower nonce equal to new confirmed nonce if needed.
+                    // Note: There will be max. 1 record to update as we don't allow messages from relayers to
+                    // overlap.
+                    entry.messages.dispatch_results >>= (new_confirmed_nonce + 1 - entry.messages.begin);
+                    entry.messages.begin = new_confirmed_nonce + 1;
+                }
+            }
+            inboundLaneNonce.last_confirmed_nonce = new_confirmed_nonce;
+        }
+        return latest_received_nonce;
+    }
+
+    // Receive new message.
+    function receive_message(Message[] memory messages) internal returns (uint256 dispatch_results) {
+        address payable relayer = msg.sender;
+        uint256 begin = inboundLaneNonce.last_delivered_nonce + 1;
+        uint256 next = begin;
+        uint256 end;
+        for (uint256 i = 0; i < messages.length; i++) {
+            Message memory message = messages[i];
+            MessageKey memory key = message.key;
+            MessagePayload memory message_payload = message.data.payload;
+            if (key.nonce < next) {
+                continue;
+            }
+            // check message nonce is correct and increment nonce for replay protection
+            require(key.nonce == next, "Lane: InvalidNonce");
+            // check message is from the correct source chain position
+            require(key.this_chain_id == bridgedChainPosition, "Lane: InvalidSourceChainId");
+            // check message delivery to the correct chain position
+            require(key.bridged_chain_id == thisChainPosition, "Lane: InvalidTargetChainId");
+            // check message delivery to the correct lane position
+            require(key.lane_id == lanePosition, "Lane: InvalidLaneID");
+            // if there are more unconfirmed messages than we may accept, reject this message
+            require(next - inboundLaneNonce.last_confirmed_nonce <= MAX_UNCONFIRMED_MESSAGES, "Lane: TooManyUnconfirmedMessages");
```

Comment:
can we assert this requirement out of the loop?

Reply:
Maybe ok, but I think it is better to keep the requirement together.